include/git2/hook.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+#ifndef INCLUDE_git_hook_h__
+#define INCLUDE_git_hook_h__
+
+#include "common.h"
+#include "types.h"
+
+/**
+ * @file git2/hook.h
+ * @brief Git hook management routines
+ * @defgroup git_hook Git hook management routines
+ * @ingroup Git
+ * @{
+ */
+GIT_BEGIN_DECL
+
+/**
+ * A helper to get the path of the repository's hooks.
+ *
+ * This will obey core.hooksPath.
+ *
+ * @param out_dir A pointer to a git_buf that will be filled with the absolute path.
+ * @param repo The repository to get the hooks location from.
+ * @return 0 on success, or an error code.
+ */
+GIT_EXTERN(int) git_hook_dir(git_buf *out_dir, git_repository *repo);
+
+typedef int (*git_hook_foreach_cb)(const char *hook_name, void *payload);
+
+/**
+ * Enumerate a repository's hooks.
+ *
+ * The callback will be called with the name of each available hook.
+ *
+ * @param repo The repository.
+ * @param callback The enumeration callback.
+ * @param payload A user-provided pointer that will be passed to the callback.
+ * @return 0 on success, or an error code.
+ */
+GIT_EXTERN(int) git_hook_foreach(
+	git_repository *repo,
+	git_hook_foreach_cb callback,
+	void *payload);
+
+/**
+ * A hook environment
+ *
+ * This structure will be provided by the library when a hook needs to
+ * be executed.
+ *
+ * - `path` is the absolute path to the hook executable.
+ * - `args` is the argument list for the hook.
+ * - `io` is either an in/out pointer to either data that must be provided to
+ *        the hook, a pointer where the executor will stash data outputted by
+ *        the hook, or both.
```

Comment:
I'm utterly at a loss about the syntax to use for documentation of typedefs. After a quick look around I can see that most use inline `/**` comments, some use a preceding `/**` comment with `*` or `-`, but it seems the doc parser ignores `*` ([case](https://github.com/libgit2/libgit2/blob/HEAD/include/git2/repository.h#L278-287) in [point](https://libgit2.github.com/libgit2/#HEAD/type/git_repository_init_options)).

In a few words, it's supposed to contain whatever input the client will have to feed the hook when the executor runs, and you can set it to some buffer on return to hand back data to whatever called the hook. It's a little fuzzy because right now the only hooks that (could) make use of this facility are `pre-push`,  `pre-receive`, `update`, `post-receive`, `post-update`, but most of these are only used on the remote side, and I don't think `libgit2` supports being used as a server (yet).

That wasn't a few words 😟.